File: server/src/plugins/runtime/node/configDb.ts
import type { Promisify } from '@camera.ui/rpc';
import type { PluginConfigStoreRPC } from '../../config-store.js';
import type { PluginStoreFile } from '../../store/pluginStoreFile.js';

export interface PluginConfigDb {
  get(key: 'config'): Record<string, any> | undefined;
  put(key: 'config', value: Record<string, any>): Promise<unknown>;
  close?(): Promise<unknown>;
}

export class LocalPluginConfigDb implements PluginConfigDb {
  constructor(private readonly store: PluginStoreFile) {}

  public get(_key: 'config'): Record<string, any> {
    return this.store.get();
  }

  public async put(_key: 'config', value: Record<string, any>): Promise<void> {
    await this.store.put(value);
  }

  public async close(): Promise<void> {
    await this.store.close();
  }
}

export class RemotePluginConfigDb implements PluginConfigDb {
  #cache: Record<string, any> = {};

  constructor(private readonly store: Promisify<PluginConfigStoreRPC>) {}

  public async init(): Promise<void> {
    this.#cache = (await this.store.get()) ?? {};
  }

  public get(_key: 'config'): Record<string, any> {
    return this.#cache;
  }

  public async put(_key: 'config', value: Record<string, any>): Promise<void> {
    this.#cache = value;
    await this.store.put(value);
  }
}

File: server/src/plugins/runtime/python/config_db.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, Protocol

if TYPE_CHECKING:
    from plugins.runtime.python.store import PluginStoreFile


class PluginConfigStoreRPC(Protocol):
    async def get(self) -> dict[str, Any] | None: ...

    async def put(self, config: dict[str, Any]) -> None: ...


class PluginConfigDb(Protocol):
    def get(self, key: Literal["config"], /) -> dict[str, Any] | None: ...

    async def put(self, key: Literal["config"], value: dict[str, Any], /) -> None: ...

    async def close(self) -> None: ...


class LocalPluginConfigDb:
    def __init__(self, store: PluginStoreFile) -> None:
        self._store = store

    def get(self, key: Literal["config"], /) -> dict[str, Any]:
        return self._store.get()

    async def put(self, key: Literal["config"], value: dict[str, Any], /) -> None:
        await self._store.put(value)

    async def close(self) -> None:
        await self._store.close()


class RemotePluginConfigDb:
    def __init__(self, store: PluginConfigStoreRPC) -> None:
        self._store = store
        self._cache: dict[str, Any] = {}

    async def init(self) -> None:
        self._cache = (await self._store.get()) or {}

    def get(self, key: Literal["config"], /) -> dict[str, Any]:
        return self._cache

    async def put(self, key: Literal["config"], value: dict[str, Any], /) -> None:
        self._cache = value
        await self._store.put(value)

    async def close(self) -> None:
        return None
